Core/clim-core/redefinition.lisp

```diff
@@ -0,0 +1,177 @@
+;;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;;;; Library General Public License for more details.
+;;;;
+;;;; You should have received a copy of the GNU Library General Public
+;;;; License along with this library; if not, write to the
+;;;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+;;;; Boston, MA  02111-1307  USA.
+
+(in-package #:clim-internals)
+
+;;; Frame class redefinition notification protocol
+;;;
+;;; This is used for notifying application frame classes after aspects
+;;; such as the class itself, the `generate-panes' method or the
+;;; command table have been redefined. The notification sequence
+;;; begins with a call to `note-frame-class-redefined' which either
+;;; directly calls `handle-frame-class-redefinition' or schedules the
+;;; call to be made when leaving a
+;;; `with-delayed-redefinition-notifications'.
+;;;
+;;; This protocol is probably most commonly initiated by re-evaluating
+;;; the `define-application-frame' form that defined the frame class
+;;; in question, but any event that causes the `reinitialize-instance'
+;;; method of the class to execute also triggers it. In the context of
+;;; `define-application-frame',
+;;; `with-delayed-redefinition-notifications' delays the
+;;; `note-frame-class-redefinition' call until after all aspects
+;;; (`generate-panes' method, command table definition, etc) have been
+;;; executed.
+
+(defgeneric handle-frame-class-redefinition (class)
+  (:method ((class t))))
+
+(defvar *delayed-redefinition-notifications*)
+
+(defun call-with-delayed-redefinition-notifications (thunk)
+  (if (boundp '*delayed-redefinition-notifications*)
+      (funcall thunk)
+      (let ((*delayed-redefinition-notifications* '()))
+        (multiple-value-prog1
+            (funcall thunk)
+          (map nil #'handle-frame-class-redefinition
+               *delayed-redefinition-notifications*)))))
+
+(defmacro with-delayed-redefinition-notifications (&body body)
+  `(call-with-delayed-redefinition-notifications (lambda () ,@body)))
+
+(defun note-frame-class-redefined (class)
+  (if (boundp '*delayed-redefinition-notifications*)
+      (pushnew class *delayed-redefinition-notifications* :test #'eq)
+      (handle-frame-class-redefinition class)))
+
+;;; `frame-class-redefined-event'
+;;;
+;;; Sent to frame instances whose class has been redefined. Causes the
+;;; following actions in the frame receiving the event:
+;;;
+;;; 1) A new command table is installed in the frame.
+;;;
+;;; 2) The current layout of the frame is reset to the first layout in
+;;;    the layout list of the frame.
+;;;
+;;; 3) Panes are regenerated (actually reinitialized where possible)
+;;;    for the potentially changed layout.
+
+(defclass frame-class-redefined-event (standard-event)
+  ((%new-layouts       :initarg :new-layouts
+                       :reader  new-layouts)
+   (%new-command-table :initarg :new-command-table
+                       :reader  new-command-table)))
```

Comment:
I have been wondering about the usefulness of that macro. Given that `event-type` is not used anywhere (except, maybe, by some clients), wouldn't it make sense to define a general `event-type` method or function that implements the procedure described in the spec instead of defining a method for each event class?

Reply:
maybe so - another approach would be having a slot in the standard-event class which is initialized to event's type and being the reader.

But until we do that, let's keep this definition consistent with the rest.

Reply:
Changed.